torch/lib/c10d/Utils.hpp

```diff
@@ -3,18 +3,124 @@
 #include <sys/socket.h>
 #include <sys/types.h>
 
-#include <cstdlib>
+#include <chrono>
 #include <cstdint>
+#include <cstdlib>
 #include <functional>
 #include <limits>
 #include <string>
 #include <system_error>
 #include <tuple>
 #include <vector>
-#include <chrono>
+
+#include <ATen/ATen.h>
+
+#include "Types.hpp"
+
+inline void hash_combine(std::size_t& seed) {}
+
+template <typename T, typename... Rest>
+inline void hash_combine(std::size_t& seed, const T& v, Rest... rest) {
+  std::hash<T> hasher;
+  seed ^= hasher(v) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+  hash_combine(seed, rest...);
+}
+
+#define MAKE_HASHABLE(type, ...)                  \
+  namespace std {                                 \
+  template <>                                     \
+  struct hash<type> {                             \
+    std::size_t operator()(const type& t) const { \
+      std::size_t ret = 0;                        \
+      hash_combine(ret, __VA_ARGS__);             \
+      return ret;                                 \
+    }                                             \
+  };                                              \
+  }
+
+// Make std::vector<T> hashable
+namespace std {
+template <typename T>
+struct hash<vector<T>> {
+  std::size_t operator()(const vector<T>& v) const {
+    std::size_t ret = v.size();
+    for (const auto& e : v) {
+      hash_combine(ret, e);
+    }
+    return ret;
+  }
+};
+} // namespace std
+
+// Make at::ArrayRef<T> hashable
+namespace std {
+template <typename T>
+struct hash<at::ArrayRef<T>> {
+  std::size_t operator()(const at::ArrayRef<T>& v) const {
+    std::size_t ret = v.size();
+    for (const auto& e : v) {
+      hash_combine(ret, e);
+    }
+    return ret;
+  }
+};
+} // namespace std
+
+MAKE_HASHABLE(::c10d::CollectiveType, static_cast<std::uint8_t>(t));
+MAKE_HASHABLE(::c10d::DeviceType, static_cast<std::uint8_t>(t));
+MAKE_HASHABLE(::c10d::ReduceOp, static_cast<std::uint8_t>(t));
 
 namespace c10d {
 
+inline void assertSameSizeAndType(const std::vector<at::Tensor>& tensors) {
+  // Ensure we have at least one tensor
+  if (tensors.size() == 0) {
+    throw std::invalid_argument("argument is empty");
+  }
+
+  // Ensure all tensors have identical type and shape
+  auto& type = tensors[0].type();
+  auto sizes = tensors[0].sizes();
+  for (auto i = 1; i < tensors.size(); i++) {
+    if (tensors[i].type() != type) {
+      throw std::invalid_argument("argument contains mixed types");
+    }
+    if (!tensors[i].sizes().equals(sizes)) {
+      throw std::invalid_argument("argument contains mixed sizes");
+    }
+  }
+}
+
+inline std::vector<std::vector<int64_t>> getSizes(
+    const std::vector<at::Tensor>& tensors) {
+  std::vector<std::vector<int64_t>> sizes(tensors.size());
+  for (auto i = 0; i < tensors.size(); i++) {
+    sizes[i] = tensors[i].sizes();
+  }
+  return sizes;
+}
```

Comment:
1. It might be better to return a vector of `ArrayRef`s to prevent forcing an allocation for every tensor.
2. Some time ago I've implemented a few utilities for mapping/filtering vectors in PyTorch C++ code. You can find it in `torch/csrc/utils/functional.h`. The thing you wrote here could then become
```cpp
fmap(tensors, [](const at::Tensor& t) { return t.sizes(); });
```
and in fact it would be even more efficient, because it makes sure to `.reserve` the memory and avoid multiple allocations. It's something that everyone always forgets about.

This also lets you easily handle functions like `getDevices`, `getDataPointers`, and any future cases where you basically want to map a simple function over a vector.

Reply:
Nice, we should pull it into aten/utils or something.

Regarding return a vector of ArrayRefs: this would only work if the tensors they are retrieved from are guaranteed to stick around. This is not the case. This vector of sizes is used in the algorithm key, which has a lifecycle that is different from the tensors for which it was first created.

Reply:
It would still work. ArrayRefs are implicitly convertible to `std::vector`s, so they would automatically allocate a copy if you were to assign them to a vector attribute, like in the algorithm key case. They would however save us overhead in all other cases.

Reply:
But it's not a very important thing, just a general note

Reply:
I'll keep it in mind. I don't think it is applicable here because this function gets a vector of sizes. For the assignment operator to work we need a vector<ArrayRef> to be convertible to vector<vector<int64_t>>, which I don't think exists.